pkg/objstore/objstore.go

```diff
@@ -437,6 +437,87 @@ func (b *metricBucket) Name() string {
 	return b.bkt.Name()
 }
 
+type prefixedBucket struct {
+	bkt    Bucket
+	prefix string
+}
+
+// NewPrefixedBucket returns a new prefixedBucket.
+func NewPrefixedBucket(bkt Bucket, prefix string) *prefixedBucket {
+	return &prefixedBucket{
+		bkt:    bkt,
+		prefix: prefix,
+	}
+}
+
+func (pbkt *prefixedBucket) WithExpectedErrs(expectedFunc IsOpFailureExpectedFunc) Bucket {
+	if ib, ok := pbkt.bkt.(InstrumentedBucket); ok {
+		return &prefixedBucket{
+			bkt:    ib.WithExpectedErrs(expectedFunc),
+			prefix: pbkt.prefix,
+		}
+	}
+	return pbkt
+}
+
+func (pbkt *prefixedBucket) ReaderWithExpectedErrs(expectedFunc IsOpFailureExpectedFunc) BucketReader {
+	return pbkt.WithExpectedErrs(expectedFunc)
+}
+
+func (pbkt *prefixedBucket) Name() string {
+	return pbkt.bkt.Name()
+}
+
+func (pbkt *prefixedBucket) Upload(ctx context.Context, name string, r io.Reader) (err error) {
+	pname := filepath.Join(pbkt.prefix, name)
+	err = pbkt.bkt.Upload(ctx, pname, r)
+	return
+}
+
+func (pbkt *prefixedBucket) Delete(ctx context.Context, name string) (err error) {
+	pname := filepath.Join(pbkt.prefix, name)
+	err = pbkt.bkt.Delete(ctx, pname)
+	return
+}
+
+func (pbkt *prefixedBucket) Close() error {
+	return pbkt.bkt.Close()
+}
+
+func (pbkt *prefixedBucket) Iter(ctx context.Context, dir string, f func(string) error) error {
+	pdir := filepath.Join(pbkt.prefix, dir)
+	if pbkt.prefix != "" {
+		return pbkt.bkt.Iter(ctx, pdir, func(s string) error {
+			return f(strings.Join(strings.Split(s, pbkt.prefix)[1:], "/"))
```

Comment:
Shouldn't be `f(strings.Join(strings.Split(s, "/")[1:], "/"))`?

Reply:
I think this is fine. but would be more sure after writing a unit test

Reply:
The units tests for Prefixed bucket Iter are running fine locally.